Notary integration

dmcgowan opened the pull request:
# Summary

Add support for trusted image distribution through the use of notary. Notary logic is done client side to allow cli users full configuration of trust and limiting access to private keys to user's machine. Trusted cli actions which involve pulling will always do a pull by digest to the daemon. Since the daemon already supports pull by digest, no changes to the daemon are needed to support the use of notary. All signatures will be done directly from the client on push. For more details on the design, see the [design document](https://docs.google.com/document/d/1Xn9kuhICOFj_P37ZWng8LeR_32NAj4Ytm42XAVWuTHw/edit#).
# UX Changes
## Flags
- `--untrusted` - Explicitly run a pull or push action as untrusted. Defaults to true in 1.7. This can be used to override the `DOCKER_NOTARY` environment variable.
## Trusted Mode
### Pull Changes
#### Pull by digest

The initial pull line will now always include the digest of the image being pulled.
#### Sequential pull all

When pulling multiple tags, the client will now sequentially call a pull by digest on every tag.
#### Full sample pull output

```
Expanded name to docker.io/dmcgowan/myapp
Pull (1 of 1): dmcgowan/myapp:latest@sha256:8d62273b15d0dd7726e5dd7e27b5a1c8bff2106557caccfa5e2f1ba6956556c8
sha256:8d62273b15d0dd7726e5dd7e27b5a1c8bff2106557caccfa5e2f1ba6956556c8: Pulling from dmcgowan/myapp
a8219747be10: Already exists 
91c95931e552: Already exists 
Digest: sha256:8d62273b15d0dd7726e5dd7e27b5a1c8bff2106557caccfa5e2f1ba6956556c8
Status: Downloaded newer image for dmcgowan/myapp@sha256:8d62273b15d0dd7726e5dd7e27b5a1c8bff2106557caccfa5e2f1ba6956556c8
Tagging dmcgowan/myapp@sha256:8d62273b15d0dd7726e5dd7e27b5a1c8bff2106557caccfa5e2f1ba6956556c8 as dmcgowan/myapp:latest
```
### Push Changes
#### Status Line

Push status line which displays digest now shows tag, digest, and size. This was used for client to be able to parse from the output what the pushed manifests were. These output lines represent the changes which will be made to the notary repository.

old

```
Digest: sha256:f2ee4bbffcce22f295f8c9a987dcc55f0987b339bf8fc0277e52560563e5c118
```

new

```
latest: digest: sha256:f2ee4bbffcce22f295f8c9a987dcc55f0987b339bf8fc0277e52560563e5c118 size: 10115
```
#### Passphrase prompt on first push of repository

When a repository is pushed for the first time the cli will be prompted for a passphrase. The passphrase will not be echoed back to the user. If metadata is near expiration time, the passphrase may be prompted again to re-sign the metadata. Subsequent pushes should not require entering a passphrase. If entering in a passphrase is not desirable, then the `notary` tool may be used directly to initialize a repository.
#### Full sample push output

```
The push refers to a repository [dmcgowan/myapp] (len: 1)
91c95931e552: Image already exists 
a8219747be10: Image already exists 
latest: digest: sha256:f2ee4bbffcce22f295f8c9a987dcc55f0987b339bf8fc0277e52560563e5c118 size: 10115
Signing and pushing TUF metadata
Expanded name to docker.io/dmcgowan/myapp
Loading trusted collection.
Creating a new private key
Passphrase:
Saving changes to Trusted Collection.
Finished initializing "docker.io/dmcgowan/myapp"
Loading trusted collection.
```
## Environment Variables
- `DOCKER_NOTARY` - boolean whether to use Notary in Docker
- `NOTARY_SERVER` - Notary server to use for all repositories
# Remaining items for code review
- [x] Authentication integration with distribution client (need to solve problem with FQDN support)
- [x] Load Certificates from ~/.docker/tls/<hostname>

NathanMcCauley committed:
Add cmdline docs for signing and verification flows

Signed-off-by: Nathan McCauley <nathan.mccauley@docker.com>

jfrazelle commented:
Can we combine this PR with https://github.com/docker/docker/pull/14453 so the conversation is not fragmented

jfrazelle commented:
Also how will these changes be effected by the refactor of the registry code, which I hope will get in before this PR if I am not mistaken right?

dmcgowan commented:
@jfrazelle I decoupled the registry refactor for the initial PR so this would not be dependent on the other. The authentication code will require using code from the registry client though. Either way we can get this in without the other.

calavera commented:
I wonder if we should put this in experimental. That way we'd have more time to polish it.

jfrazelle commented:
is this the same vendored uuid library we were having issues with

dmcgowan commented:
I think it is the same library, will submit a PR to notary to use the distribution version

dmcgowan commented:
Updated uuid library

icecrime commented on vendor/src/github.com/tent/canonical-json-go/encode.go:
```diff
@@ -0,0 +1,620 @@
+// Copyright 2010 The Go Authors.  All rights reserved.
```
Are we really going to depend on a fork on the standard `encoding/json` for a diff which I personally can't find because the file content was reordered?

icecrime commented on hack/vendor.sh:
```diff
@@ -36,6 +36,11 @@ clone git github.com/hashicorp/consul v0.5.2
 
+clone git github.com/docker/notary 17af21f00c9fa4dc960ca8244f230ca082bd8745
+clone git github.com/endophage/gotuf ab4ba80203ffa5bfd742e6891bd28bfbf43a9453
+clone git github.com/tent/canonical-json-go 96e4ba3a7613a1216cbd1badca4efe382adea337
```
As mentioned elsewhere, this is a super weird dependency oO

endophage replied:
Reliable hashes of json data (for validation) require canonical serialization.

icecrime replied:
Alright, that it still a super shitty dependency

icecrime commented on hack/vendor.sh:
```diff
@@ -36,6 +36,11 @@ clone git github.com/hashicorp/consul v0.5.2
+clone git github.com/docker/notary 17af21f00c9fa4dc960ca8244f230ca082bd8745
+clone git github.com/endophage/gotuf ab4ba80203ffa5bfd742e6891bd28bfbf43a9453
+clone git github.com/tent/canonical-json-go 96e4ba3a7613a1216cbd1badca4efe382adea337
+clone git github.com/agl/ed25519 d2b94fd789ea21d12fac1a4443dd3a3f79cda72c
```
Isn't the signing done server side? Why do we need this in Docker?

dmcgowan replied:
We are currently not signing using ed25519 but may need to verify these signatures.

icecrime commented on api/client/push.go:
```diff
@@ -42,6 +50,136 @@ func (cli *DockerCli) CmdPush(args ...string) error {
+	} else {
+		return err
+	}
+	return repo.Publish(passFunc)
```
Can't we avoid writing new code that introduces 130+ LoC and 5 level of depth to a single function?

jfrazelle commented on hack/vendor.sh:
```diff
@@ -36,6 +36,11 @@ clone git github.com/hashicorp/consul v0.5.2
 clone git github.com/docker/distribution 419bbc2da637d9b2a812be78ef8436df7caac70d
 
+clone git github.com/docker/notary 17af21f00c9fa4dc960ca8244f230ca082bd8745
+clone git github.com/endophage/gotuf ab4ba80203ffa5bfd742e6891bd28bfbf43a9453
```
is this going to move under docker

dmcgowan replied:
@endophage do you plan on moving this?

endophage replied:
No, this goal is to work with the flynn developers to get gotuf upstream.

jfrazelle commented:
some tests would be nice, even just unit tests, but integration maybe at least one

jfrazelle commented on api/client/create.go:
```diff
@@ -94,20 +93,37 @@ func (cli *DockerCli) createContainer(config *runconfig.Config, hostConfig *runc
+	ref := registry.ParseReference(tag)
+	var trustedRef registry.Reference
+
+	if isTrusted(repo) && !ref.IsTrusted() {
```
shouldnt there be a TODO for when isTrusted(repo) is false we will eventually want to explode and die with 80 bajillion warnings

jfrazelle commented on api/client/create.go:
```diff
@@ -94,20 +93,37 @@ func (cli *DockerCli) createContainer(config *runconfig.Config, hostConfig *runc
 		if err = cli.pullImageCustomOut(config.Image, cli.err); err != nil {
 			return nil, err
 		}
+		if trustedRef != nil && !ref.IsTrusted() {
```
this is a double negative and then it looks like we are tagging it as trusted.... ok...

dmcgowan replied:
I need to add a comment here. This is adding a tag from the an untrusted reference `ref` to a trusted reference `trustedRef`, this would be the case where you asked for `ubuntu:latest` but we pulled `ubunut@sha256...`, then will add the tag for `ubuntu:latest` pointing at `ubunut@sha256...`. This could possibly be done as well as pull `ubuntu:latest@sha256...` but would require additional code to support in the daemon.

jfrazelle commented on api/client/pull.go:
```diff
@@ -17,24 +16,29 @@ import (
-	v.Set("fromImage", newRemote)
+	ref := registry.ParseReference(tag)
+
+	if isTrusted("") && !ref.IsTrusted() {
```
why? are we not doing this server side

dmcgowan replied:
referring to which part of the statement? The `!ref.IsTrusted()` sends pull by digests down the existing code path even when trust is enabled. It could go the trusted path as well, would just require a check similar to create, only create a tag if asked for untrusted tag `ubuntu:latest` but pulled trusted `ubuntu@sha256...`.

jfrazelle replied:
this was my b i thought it was also being done in a pull file

On Mon, Jul 13, 2015 at 4:44 PM, Derek McGowan <notifications@github.com>
wrote:

> In api/client/pull.go
> <https://github.com/docker/docker/pull/14546#discussion_r34524690>:
>
> >  		return fmt.Errorf("tag can't be used with --all-tags/-a")
> >  	}
> >
> > -	v.Set("fromImage", newRemote)
> > +	ref := registry.ParseReference(tag)
> > +
> > +	if isTrusted("") && !ref.IsTrusted() {
>
> referring to which part of the statement? The !ref.IsTrusted() sends pull
> by digests down the existing code path even when trust is enabled. It could
> go the trusted path as well, would just require a check similar to create,
> only create a tag if asked for untrusted tag ubuntu:latest but pulled
> trusted ubuntu@sha256....
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/docker/docker/pull/14546/files#r34524690>.
>

jfrazelle commented on api/client/push.go:
```diff
@@ -42,6 +50,136 @@ func (cli *DockerCli) CmdPush(args ...string) error {
+	streamOut := cli.out
+	targets := []target{}
+	trustedPush := isTrusted(remote)
+	if trustedPush {
```
1. does this really need its own variable

2. can we not do the huge nest

jfrazelle commented on api/client/push.go:
```diff
@@ -42,6 +50,136 @@ func (cli *DockerCli) CmdPush(args ...string) error {
+	if err != nil {
+		return err
+	}
+	if !trustedPush {
```
ya this can def be done in a different order and avoid the nest

dmcgowan replied:
I'll break this out like was done with pull

jfrazelle commented on api/client/push.go:
```diff
@@ -42,6 +50,136 @@ func (cli *DockerCli) CmdPush(args ...string) error {
+		return nil
+	}
+
+	if tag == "" {
```
does this not mean tag is latest? why would there never be a tag

dmcgowan replied:
Push has the confusing behavior of treating `tag == ""` as push all tags while pull treats it as pull latest.

jfrazelle commented on api/client/push.go:
```diff
@@ -42,6 +50,136 @@ func (cli *DockerCli) CmdPush(args ...string) error {
+		if err != nil {
+			return err
+		}
+		if rootKey == "" {
```
i just feel like some of these ifs can be avoided with a different implementation

jfrazelle commented on api/client/trust.go:
```diff
@@ -0,0 +1,323 @@
+func isTrusted(name string) bool {
+	// TODO(dmcgowan): whitelist specific names to be untrusted
+	// TODO(dmcgowan): canonicalize name before checking whitelist
+	return !untrusted
```
isTrusted: return the opposite of untrusted....

dmcgowan replied:
This was flipped after initial implementation from `trusted` to `untrusted`. The function could probably be named better.

jfrazelle replied:
why cant we just name the bool trusted

On Mon, Jul 13, 2015 at 4:33 PM, Derek McGowan <notifications@github.com>
wrote:

> In api/client/trust.go
> <https://github.com/docker/docker/pull/14546#discussion_r34523973>:
>
> > +	var trusted bool
> > +	if e := os.Getenv("DOCKER_NOTARY"); e != "" {
> > +		var err error
> > +		trusted, err = strconv.ParseBool(e)
> > +		if err != nil {
> > +			// treat any other value as true
> > +			trusted = true
> > +		}
> > +	}
> > +	fs.BoolVar(&untrusted, []string{"-untrusted"}, !trusted, "Use or update trusted repositories on push and pull")
> > +}
> > +
> > +func isTrusted(name string) bool {
> > +	// TODO(dmcgowan): whitelist specific names to be untrusted
> > +	// TODO(dmcgowan): canonicalize name before checking whitelist
> > +	return !untrusted
>
> This was flipped after initial implementation from trusted to untrusted.
> The function could probably be named better.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/docker/docker/pull/14546/files#r34523973>.
>

dmcgowan replied:
This variable is what is set by the flag set.

jfrazelle replied:
bleh grosss... idk how to fix

jfrazelle replied:
well we have the trusted var above... this is just kinda a lot of !var lol but i dont know of a solution so i digress

tiborvass replied:
@dmcgowan what would be examples of automatically untrusted names? should it be related to insecure registries (aka resolves to 127/8 or was configured)

dmcgowan replied:
I think when I flipped the logic I also changed this comment. I was thinking initially we may have a whitelist to automatically trust images such as "docker.io/library", but you raise a good point that images within a "local" namespace could be treated as untrusted.

bfirsh commented:
Just to echo @dave-tucker's comments in #14453 that the `--untrusted` flag is a bit odd. `--verify=true` and a user passing `--verify=false` seems clearer.

I'd also agree "trusted" is quite a loaded word. (Trusted by who?) The thing it's actually doing is signing and verifying the signatures.

tiborvass commented:
@bfirsh hmm, I'd argue it's better to have a negatively connotated flag such as `untrusted` or `noverify`, or `insecure` instead of playing with the `=false`. It's secure by default, and if you don't want it to be secure, you add the naughty flag without `=` symbol.

Indeed it might not be pretty when it's not on by default, you would have that double negative `untrusted=false`, but that's not gonna stay like it. In 1.9 it will be exactly like we want it: trusted by default, otherwise user needs to pass `--untrusted`.

As for the choice of the word itself I don't know what's best, but I don't feel the problem with the word *trust*.

jfrazelle commented:
How is this not minus more lines? Can't we get rid of all the crap server side now that validates

calavera commented:
@dmcgowan it looks like there are some formatting issues in the build:

```
---> Making bundle: validate-vet (in bundles/1.8.0-dev/validate-vet)
api/client/trust.go:284: no formatting directive in Fprintf call
exit status 1
```

dmcgowan committed:
Use notary library for trusted image fetch and signing

Add a trusted flag to force the cli to resolve a tag into a digest via the notary trust library and pull by digest.
On push the flag the trust flag will indicate the digest and size of a manifest should be signed and push to a notary server.
If a tag is given, the cli will resolve the tag into a digest and pull by digest.
After pulling, if a tag is given the cli makes a request to tag the image.

Use certificate directory for notary requests

Read certificates using same logic used by daemon for registry requests.

Catch JSON syntax errors from Notary client

When an uncaught error occurs in Notary it may show up in Docker as a JSON syntax error, causing a confusing error message to the user.
Provide a generic error when a JSON syntax error occurs.

Catch expiration errors and wrap in additional context.

Signed-off-by: Derek McGowan <derek@mcgstyle.net> (github: dmcgowan)

dmcgowan commented:
@jfrazelle the existing signature verification is not related to this change. Any changes to server side verification should be done in its own change.

GordonTheTurtle commented:
Please sign your commits following these rules:
https://github.com/docker/docker/blob/master/CONTRIBUTING.md#sign-your-work
The easiest way to do this is to amend the last commit:
~~~console
$ git clone -b "trusted-notary-integration" git@github.com:dmcgowan/docker.git somewhere
$ cd somewhere
$ git rebase -i HEAD~4
editor opens
change each 'pick' to 'edit'
save the file and quit
$ git commit --amend -s --no-edit
$ git rebase --continue # and repeat the amend for each commit
$ git push -f
~~~

Ammending updates the existing PR. You **DO NOT** need to open a new one.

calavera commented:
@dmcgowan it looks like there is something missing:

```
api/client/trust.go:485: undefined: trustDirectory
```

tiborvass commented on api/client/trust.go:
```diff
@@ -0,0 +1,525 @@
+		scanner.Split(ScanInput)
+		// New splitter
+		for scanner.Scan() {
+			line := scanner.Text()
```
use `Bytes()` instead, so that you can use `FindSubmatch` below

tiborvass commented on api/client/trust.go:
```diff
@@ -0,0 +1,525 @@
+	}
+
+	// Close stream channel to finish target parsing
+	if err := streamOut.Close(); err != nil {
```
what if the function returns earlier (e.g. the push call fails) ?

dmcgowan replied:
Hmm, guessing the scanner would hang since I don't think an EOF will be sent. Could always close before checking the previous error.

tiborvass commented on api/client/trust.go:
```diff
@@ -0,0 +1,525 @@
+	}
+
+	err = repo.Publish(passFunc)
+	if _, ok := err.(*client.ErrRepoNotInitialized); ok {
```
small nit: you could just do `if ...; !ok { return }` and then not have the rest indented so much.

tiborvass commented on api/client/trust.go:
```diff
@@ -0,0 +1,525 @@
+			rootKey = keys[0]
+			fmt.Fprintf(cli.out, "Using private key %q\n", rootKey)
+		}
+		passphrase, err := passFunc()
```
i find this weird to call passFunc when it's a function meant to be called by notary iirc. Is this code some boilerplate that could go to notary?

dmcgowan replied:
The whole function passing is in my opinion broken, there should not be two different locations which would prompt for a passphrase like this. However the getting the passphrase does need to happen in a  uniform way and that is the role of the CLI. If notary was going to change it should allow the client to own the passphrase flow and only require passing in a crypto service, I have asked for this :angel:

endophage replied:
The requirement I have heard has been:

Only ask for the password when it's required (because it's required infrequently so asking for it every time is annoying).

To achieve that requirement there are 2 options:

1. Make the caller of of the NotaryClient functions more aware of what's going on. - This has been deemed undesirable because it will leak more Notary/TUF related things up the call stack.
2. Enable NotaryClient to retrieve the password if and when it needs it.

With that in mind, what is the most basic thing we need? It's a function that takes no arguments, and returns a password (and maybe an error). That function could be defined on the CryptoService, it could be defined on some new Authenticator struct (an idea Stephen raised last week), but however it is defined, that specific function can still be passed in where it is required.

This implementation is going to be the most stable way to handle this while still leaving open a plethora of options for improving exactly how and when a password is retrieved.

jfrazelle commented:
There are still literally no tests with this PR.

Spin up a notary server in the container in you have to, we already do this with registry it shouldn't be hard.

One test should atleast test the keys are setup properly and then we should also test push and pull.

Something referred to as "trust" should include tests, otherwise how can I trust it.

jfrazelle commented:
Also I can't personally test this because it looks like the infrastructure is not in place yet... are we seriously going down this road again

jfrazelle commented on api/client/trust.go:
```diff
@@ -0,0 +1,534 @@
+		return s
+	}
+	if index.Official {
+		return "https://notary-stage.docker.io"
```
I take it these images have not been pushed and/or the infrastructure is not in place yet, is there an estimated date for this. I really am not a fan of going down this road it is super stressful for all when the stuff is not in place a day before the release and it has not been properly tested.

dmcgowan commented:
Pushed change to address code comments.

There are two outstanding issues though to satisfy @jfrazelle's comments. Integration-cli tests are in progress and will be added soon. The notary staging service should be able to be used for testing, if there are issue please report. The production endpoint will be up early next week and this PR will get updated then. I do not expect a merge before testing has been done on the production endpoint.

diogomonica commented on api/client/trust.go:
```diff
@@ -0,0 +1,456 @@
+	targets := <-targetChan
+
+	if tag == "" {
+		fmt.Fprintf(cli.out, "No tag specified, skipping TUF metadata push\n")
```
This should be Notary metadata. Also currently we don't seem to assume latest when no tag gets specified:

```sh
➜  notary git:(cooking-the-tofu-to-support-rotations) ✗ docker push diogomonica/alpine
The push refers to a repository [docker.io/diogomonica/alpine] (len: 2)
a9539b34a6ab: Image already exists 
b3dbab3810fc: Image already exists 
latest: digest: sha256:d149ab53f8718e987c3a3024bb8aa0e2caadf6c0328f1d9d850b2a2a67f2819a size: 3355
No tag specified, skipping TUF metadata push
```

dmcgowan replied:
Can update to say "Notary metadata".  As for whether we default to latest when not specified, this behavior is kind of inconsistent. If you were to pull without a tag, it will default to latest, but if you push without a tag, it will push all tags for a repository. If this check is not in place on push, it will sign all tags. If this is the desired behavior, we can remove it. My impression was that we wanted the user to be more explicit about what is signed. If we want to add UX to make the user confirm the signature, we can discuss that as well.

endophage commented on api/client/trust.go:
```diff
@@ -0,0 +1,456 @@
+		return passphrase, nil
+	}
+
+	err = repo.Publish(passFunc)
```
Any reason not to just pass `cli.getPassphrase` here as opposed to wrapping it in `passFunc`?

jfrazelle replied:
If we have to have it I like it in the function

> On Jul 18, 2015, at 23:15, David Lawrence <notifications@github.com> wrote:
> 
> In api/client/trust.go:
> 
> > +			Length: int64(target.size),
> > +		}
> > +		if err := repo.AddTarget(t); err != nil {
> > +			return err
> > +		}
> > +	}
> > +
> > +	passFunc := func() (string, error) {
> > +		passphrase, err := cli.getPassphrase()
> > +		if err != nil {
> > +			return "", err
> > +		}
> > +		return passphrase, nil
> > +	}
> > +
> > +	err = repo.Publish(passFunc)
> Any reason not to just pass cli.getPassphrase here as opposed to wrapping it in passFunc?
> 
> —
> Reply to this email directly or view it on GitHub.
>

endophage replied:
It is in a function, `cli.getPassphrase`, the `passFunc` has an identical signature. `passFunc` could even just be rewritten as `return cli.getPassphrase()`, which makes it seem like an even less purposeful wrapper.

jfrazelle replied:
Both are pretty gross so it's relative 

> On Jul 19, 2015, at 00:20, David Lawrence <notifications@github.com> wrote:
> 
> In api/client/trust.go:
> 
> > +			Length: int64(target.size),
> > +		}
> > +		if err := repo.AddTarget(t); err != nil {
> > +			return err
> > +		}
> > +	}
> > +
> > +	passFunc := func() (string, error) {
> > +		passphrase, err := cli.getPassphrase()
> > +		if err != nil {
> > +			return "", err
> > +		}
> > +		return passphrase, nil
> > +	}
> > +
> > +	err = repo.Publish(passFunc)
> It is in a function, cli.getPassphrase, the passFunc has an identical signature. passFunc could even just be rewritten as return cli.getPassphrase(), which makes it seem like an even less purposeful wrapper.
> 
> —
> Reply to this email directly or view it on GitHub.
>

icecrime commented:
I'm -1 against that last commit, as I already mentioned in the past to @NathanMcCauley. I find `FROM` rewriting ugly, and I don't want to end up with Compose (or any other element of the toolbox for that matter) implement a hackish Dockerfile rewrite to accommodate for the fact that trust is implement client-side. If you want trust enabled daemon initiated actions, reconsider that design choice.

icecrime commented:
Also, would _you_ merge in untested code that says that?

```
// TODO: this is awful, so implement a client-side builder that uses the
// existing pull logic.
```

jfrazelle commented:
I offered to help them fix this... I don't like a lot of it either, I still have to work on my talk today but I hate when we get PRs like this last minute and it always seems like it happens wrt the registry code base. I don't know what to say other than I'm disappointed.

> On Jul 19, 2015, at 11:08, Arnaud Porterie <notifications@github.com> wrote:
> 
> Also, would you merge in untested code that says that?
> 
> // TODO: this is awful, so implement a client-side builder that uses the
> // existing pull logic.
> —
> Reply to this email directly or view it on GitHub.
>

NathanMcCauley commented:
@arnaud @jfrazelle I also find FROM rewriting ugly. You could call it a transitionary change until we have a client side builder. You could call also call it a hack :) Once we have a client side builder this will go away. Compose or other API clients will not have to do Dockerfile rewriting until they have trust, since that's the only purpose of it. They'll be able to implement client side building and trust at the same time.

jfrazelle commented:
Docker shouldn't be the code base with hacks when it is the one being used in production environments

> On Jul 19, 2015, at 12:22, Nathan McCauley <notifications@github.com> wrote:
> 
> @arnaud @jfrazelle I also find FROM rewriting ugly. You could call it a transitionary change until we have a client side builder. You could call also call it a hack :) Once we have a client side builder this will go away. Compose or other API clients will not have to do Dockerfile rewriting until they have trust, since that's the only purpose of it. They'll be able to implement client side building and trust at the same time.
> 
> —
> Reply to this email directly or view it on GitHub.
>

jlhawn commented:
If trusted pushes and pulls are considered 'experimental' in 1.8 anyway, would it be acceptable to limit it to just `docker push` and `docker pull` then? Since I'm the one who wrote the `this is awful` comment, you can trust that I really don't like it either as it essentially introduces a miniature Dockerfile parser/rewriter *while uploading context* :grimacing: which leads to some really confusing error messages when the base image is untrusted.

Perhaps the compromise could be that if users want to use experimental trusted pulls for `docker build` in 1.8 then they can *consciously* do a trusted pull of the base image and run `docker build` without the `--pull` option. This essentially makes builds with trusted base images a 2-step process from the user's perspective but probably one they are willing to accept while the feature is still considered experimental.

dmcgowan committed:
Add integration cli trust tests

Added notary server to docker base image.
Created trust suite which runs trust server for running trusted commands.

Signed-off-by: Derek McGowan <derek@mcgstyle.net> (github: dmcgowan)

cpuguy83 commented on registry/config.go:
```diff
@@ -32,6 +32,7 @@ const (
 	REGISTRYSERVER = DEFAULT_V2_REGISTRY
 	INDEXSERVER    = DEFAULT_V1_REGISTRY + "/v1/"
 	INDEXNAME      = "docker.io"
+	NOTARYSERVER   = "https://notary-stage.docker.io"
```
In order to merge, this should be pointing at the real/prod notary server.
You can control what backend this is going to via DNS.
I don't see why we should have just one more thing that needs to change in the code here.

cpuguy83 commented:
There are a _lot_ of `TODO` comments in this.
I assume these will be cleaned up before it's ready for merge?

calavera commented on api/client/trust.go:
```diff
@@ -0,0 +1,460 @@
+}
+
+func isTrusted(name string) bool {
+	// TODO(dmcgowan): false if name is hex image id
```
can we not define an argument if we're not going to use it right away? The TODO list sounds pretty reasonable, but the argument doesn't make sense util we tackle the list.

calavera commented:
I just noticed that we don't have much documentation about what's going on here besides adding the flag to the correct documents. Is there any work in progress to provide a user guide about notary and signed images?

diogomonica commented:
Yes, absolutely. We are focusing all of our time in ensuring quality of the code-base, but will be working on step-by-step instructions on how to enable and use trust in docker, and how to use notary for advanced functionality.

calavera commented on registry/registry.go:
```diff
@@ -61,6 +65,54 @@ func hasFile(files []os.FileInfo, name string) bool {
+// ReadCertsDirectory reads the directory for TLS certificates
+// including roots and certificate pairs and updates the
+// provided TLS configuration.
+func ReadCertsDirectory(tlsConfig *tls.Config, directory string) error {
```
can we use https://github.com/docker/docker/blob/master/pkg/tlsconfig/config.go#L83 to generate the TLS config rather than doing it manually here?

dmcgowan replied:
There is some mismatch with the logic to create TLS configurations for the docker cli to daemon and daemon to the registry. The config for the docker cli passes in the root and client certificates information, while the daemon attempts to interpret the names of the files in `/etc/docker/certs.d/<hostname>`.  This logic could certainly be moved to tlsconfig, but then there would be two different methods for creating TLS client configs inside that package. This change just pulls it out from inside a registry function and exposes it, we could move it completely out of registry as well. WDYT?

calavera commented on registry/registry.go:
```diff
@@ -61,6 +65,54 @@ func hasFile(files []os.FileInfo, name string) bool {
+// including roots and certificate pairs and updates the
+// provided TLS configuration.
+func ReadCertsDirectory(tlsConfig *tls.Config, directory string) error {
+	fs, err := ioutil.ReadDir(directory)
```
can we use `filepath.Walk` rather than reading the whole directory?

dmcgowan replied:
It does appear that the rest of the file info is not even used. `Walk` is not ideal since the current logic makes an extra check within the list to see if both files for a key pair exist. More efficient would just be using the underlying directory file...

```
	d, err := os.Open(directory)
	if err != nil {
		return err
	}
	fs, err := d.Readdirnames(-1)
	d.Close()
	if err != nil {
		return err
	}
```

calavera commented:
@NathanMcCauley, @dmcgowan What do you think about this alternate implementation to change the `FROM` reference for `docker build`?

1. We read the `Dockerfile` file as soon as we have the path (before creating the tar context).
2. We get the trusted digest for that image.
3. We write a temporal dockerfile with the new digest.
4. We create the tar context pointing to the temporal dockerfile.

This will prevent us from generating any tar file whatsoever if there is no trusted digest for that image and, obviously, will prevent us from doing the taring twice like the current implementation.

calavera commented:
The problem with my implementation is that the new Dockerfile must always be inside the absolute root context, so we might need to create a hidden directory there, something like `.docker-trust`?

https://github.com/dmcgowan/docker/blob/b798c6b483ec82e6858f0324c382b66b72e64334/api/client/build.go#L151

jlhawn commented:
There are just too many ways to specify the build context:

1. A tar archive read from `stdin`
2. A Dockerfile read from `stdin` (I guess if you're build doesn't require `COPY` commands?)
3. A remote non-git URL that is sent to the daemon to download (the client never sees this context and has no way to modify the Dockerfile unless it downloads the remote context first).
4. A git URL which is cloned to a local temporary directory.
5. A local directory.

Maybe the best way to do it would be to gather all of these different ways of specifying context to a local (maybe temporary) directory.

1. tar archive read from `stdin`:
    - extract it to a local temporary directory.
    - Dockerfile will be at `tempdir/Dockerfile`
2. Dockerfile read from stdin:
    - write it to `tempdir/Dockerfile`
3. Remote URL context
    - download it to tempdir.
4. A git URL
    - already handled today by cloning to a temporary directory.
5. An existing local directory
    - this is the common case.

Now we have a local context directory. Now we can re-write the Dockerfile to something like `contextdir/.Dockerfile-trust` like @calavera suggested above. Now we should be able to handle build context upload like the local directory case with `-f` equal to `.Dockerfile-trust`.

That name might collide so it would be safer to use `ioutil.Tempfile(contextDir, ".Dockerfile-trust-")` to generate a unique/random suffix.

Commit:
Refactor api/client/build.go

Separated preparation of context and Dockerfile for
the various different methods of specifying them.

Docker-DCO-1.1-Signed-off-by: Josh Hawn <josh.hawn@docker.com> (github: jlhawn)

jfrazelle commented:
Hmmm I don't like the idea of a hidden for for rewrites this has gone too far

> On Jul 20, 2015, at 13:07, David Calavera <notifications@github.com> wrote:
> 
> The problem with my implementation is that the new Dockerfile must always be inside the absolute root context, so we might need to create a hidden directory there, something like .docker-trust?
> 
> https://github.com/dmcgowan/docker/blob/b798c6b483ec82e6858f0324c382b66b72e64334/api/client/build.go#L151
> 
> —
> Reply to this email directly or view it on GitHub.
>

tiborvass commented on integration-cli/requirements.go:
```diff
@@ -74,6 +74,16 @@ var (
+		func() bool {
+			// for now notary binary is built only if we're running inside
+			// container through `make test`. Figure that out by testing if
+			// registry binary is in PATH.
```
bad comment, should be `if notary binary is in PATH`.

dmcgowan replied:
fail, I thought I erased the evidence of copying the registry requirement :smile:, it is fixed now

Commit:
Add notary integration to `docker build`

The Dockerfile is rewritten with images references on FROM
instructions resolved to trusted digests. The rewritten Dockerfile
is swapped with the original one during context upload.

Docker-DCO-1.1-Signed-off-by: Josh Hawn <josh.hawn@docker.com> (github: jlhawn)

aaronlehmann commented on api/client/trust.go:
```diff
@@ -0,0 +1,437 @@
+	if s := os.Getenv("DOCKER_TRUST_SERVER"); s != "" {
+		if !strings.HasPrefix(s, "http") {
+			return "https://" + s
+		}
```
I think this should be an explicit check for http:// or https://. Otherwise, wouldn't a DOCKER_TRUST_SERVER value like http-mirror.mydomain.com cause problems?

tiborvass replied:
+1

endophage replied:
Given we're using TOFU, we might even want to only accept https:// URLs. @NathanMcCauley @diogomonica thoughts?

jfrazelle replied:
no http, imo

calavera replied:
super in favor of https only.

diogomonica replied:
This was actually our intention from day one.

Since we are using TOFU, we have to enforce docker-cli <-> notary-server communication using HTTPS, or a MiTM could very easily server a inject a fake certificate the first time a user connects.

This way we are using the public PKI to protect the TOFU (trust on first use). We're calling it: TOFUS!

calavera replied:
you had me with the name already

icecrime replied:
This thread is confusing me: are we addressing @aaronlehmann initial comment? I think he has a point.

icecrime replied:
Also [`url.URL`](https://golang.org/pkg/net/url/#URL) is convenient for URL manipulation (such as substituting scheme), but that's probably simple enough not to bother.

diogomonica committed:
Add more integration tests for trusted push and pull

Signed-off-by: Diogo Monica <diogo@docker.com>

jfrazelle commented on api/client/build.go:
```diff
@@ -283,10 +247,8 @@ func (cli *DockerCli) CmdBuild(args ...string) error {
 	headers.Add("X-Registry-Config", base64.URLEncoding.EncodeToString(buf))
+	headers.Set("Content-Type", "application/tar")
 
-	if context != nil {
```
why did we get rid of this check

jfrazelle replied:
oh i see line 157

jfrazelle commented on api/client/build.go:
```diff
@@ -315,3 +277,295 @@ func (cli *DockerCli) CmdBuild(args ...string) error {
+// replaces the entry with the given Dockerfile name with the contents of the
+// new Dockerfile. Returns a new tar archive stream with the replaced
+// Dockerfile.
+func replaceDockerfileTarWrapper(inputTarStream io.ReadCloser, newDockerfile *trustedDockerfile, dockerfileName string) io.ReadCloser {
```
ugh I still dont like the idea of this but whatever i guess

jfrazelle commented on api/client/build.go:
```diff
@@ -315,3 +277,295 @@ func (cli *DockerCli) CmdBuild(args ...string) error {
+			// Close the tempfile if there was an error during Notary lookups.
+			// Otherwise the caller should close it.
+			tempFile.Close()
+			os.Remove(tempFile.Name())
```
this should be outside the err i believe

jlhawn replied:
We want to make sure that the file is removed during this function *only* if there's an error (we don't want to close it before the caller has a chance to use it). When the caller calls the `Close()` method (defined on the custom struct type above) it will then close and remove this temp file.

jlhawn replied:
It's probably a good idea to take your suggestion from below (declare a var with that custom struct type just before this `defer`). Then this `defer` can just call the `Close()` method on that variable. @dmcgowan could you make this change? pwetty pweez :purple_heart:

jfrazelle commented on api/client/build.go:
```diff
@@ -315,3 +277,295 @@ func (cli *DockerCli) CmdBuild(args ...string) error {
+
+	tempFile.Seek(0, os.SEEK_SET)
+
+	return &trustedDockerfile{
```
not really a fan of this on the same line as the return perhaps we can declare it then return... i realize this is more a nit so feel free to ignore

calavera replied:
we have these kind of returns in other places, I think it's fine.

jfrazelle commented on api/client/build.go:
```diff
@@ -315,3 +277,295 @@ func (cli *DockerCli) CmdBuild(args ...string) error {
 	}
 	return err
 }
+
```
All this is a lot of code to be added _just_ for the dockerfile from rewrite, which will be obsolete after client side builder anyways, it just feels kinda gross tbh sorry

jlhawn replied:
The following 6 functions should actually make implementing the client side builder much easier because they handle the work of preparing the context directory and locating the Dockerfile. Those will not be going away once we have a client-side builder. All of that refactoring work was done in a separate commit. It was necessary to do this now so that we get then have an obvious point in the `CmdBuild` method above to perform what we need to do now with the Dockerfile.

The second commit adds the two functions that are at the end of this file (the one that parses and rewrites the `FROM` lines in a Dockerfile, and the one that modifies the tar stream). These are the ones that are *just* for this PR/feature and will be deleted once we have a client-side builder.

jfrazelle replied:
ok youve sold me

jfrazelle replied:
thanks for explaining :)

jfrazelle commented on api/client/trust.go:
```diff
@@ -0,0 +1,437 @@
+		if alias == "root" && len(rootEnv) >= 8 {
+			return rootEnv, false, nil
+		}
+		if alias == "targets" && len(targetEnv) >= 8 {
```
can these be a switch

endophage replied:
I don't think there's a way to formulate a switch that wouldn't make this uglier. You'd end up with something like the following:

```go
switch alias {
case "root":
    if len(rootEnv) >= 8 {
        return rootEnv, false, nil
    }
// and so on
```

jfrazelle commented on api/client/trust.go:
```diff
@@ -0,0 +1,437 @@
+func notaryError(err error) error {
+	if _, ok := err.(*json.SyntaxError); ok {
+		logrus.Debugf("Notary syntax error: %s", err)
+		return errors.New("no trust data available for repository")
```
so this could mean nothing else? what if there was like a network disrupt or some shit halfway through

calavera replied:
agreed with @jfrazelle, we should add the original error to the returned one, or return the original one directly. It might also be useful to turn that `Debugf` into and `Errorf`

endophage replied:
This PR catches some of the problems earlier so we never get to the JSON parsing if another error is detected.

https://github.com/endophage/gotuf/pull/29/files

jfrazelle commented on api/client/trust.go:
```diff
@@ -0,0 +1,437 @@
+			if matches := targetRegexp.FindSubmatch(line); len(matches) == 4 {
+				dgst, err := digest.ParseDigest(string(matches[2]))
+				if err != nil {
+					fmt.Fprintf(in, "Bad digest value %q, ignoring\n", string(matches[2]))
```
why are these being ignored, could use a comment or something

dmcgowan replied:
Going to change these to debug log lines and will add a comment. If this case is reached it is likely because either the input was garbled or the regexp picked up an invalid line.

jfrazelle commented on api/client/trust.go:
```diff
@@ -0,0 +1,437 @@
+	if err := streamOut.Close(); err != nil {
+		return err
+	}
+	// Check error from request
```
i think this is obvious ;)

dmcgowan replied:
Perhaps, I added because there is a scoped `err` right above it which is kind of a readability fail, could just name the above error differently, a `:` is so easy to miss.

jfrazelle replied:
ah i see nice

jfrazelle commented on docs/reference/commandline/create.md:
```diff
@@ -61,6 +61,7 @@ Creates a new container.
       --restart="no"             Restart policy (no, on-failure[:max-retry], always)
       --security-opt=[]          Security options
       -t, --tty=false            Allocate a pseudo-TTY
+      --untrusted                Skip image verification
```
should have =false

jfrazelle commented on docs/reference/commandline/pull.md:
```diff
@@ -16,6 +16,7 @@ weight=1
     Pull an image or a repository from the registry
 
       -a, --all-tags=false    Download all tagged images in the repository
+      --untrusted             Skip image verification
```
should have =true

jfrazelle commented on docs/reference/commandline/push.md:
```diff
@@ -15,5 +15,7 @@ weight=1
 
     Push an image or a repository to the registry
 
+    --untrusted                Skip image signing
```
should have =true

jfrazelle commented on docs/reference/commandline/run.md:
```diff
@@ -65,6 +65,7 @@ weight=1
       -t, --tty=false            Allocate a pseudo-TTY
       -u, --user=""              Username or UID (format: <name|uid>[:<group|gid>])
       --ulimit=[]                Ulimit options
+      --untrusted                Skip image verification
```
same

jfrazelle commented on integration-cli/trust_server.go:
```diff
@@ -0,0 +1,110 @@
+		"DOCKER_TRUST=1",
+		fmt.Sprintf("DOCKER_TRUST_SERVER=%s", s.not.address()),
+		"DOCKER_TRUST_ROOT_PASSPHRASE=12345678",
+		"DOCKER_TRUST_TARGET_PASSPHRASE=12345678",
```
maybe these should be randomized per run for a better test of randomness and that things work ;)

jfrazelle replied:
we could catch some weird outlier like when its all aaaaaaaaaaaa's everything explodes and catches on fire hehe

endophage replied:
Should have separate fuzzing test for that. Tests should fail for deterministic reasons. If a fuzzing test ever causes a failure, I'd expect the problematic value to be captured and a new test added specifically for that value to ensure no regressions in the future.

jfrazelle replied:
I like random, you'd be surprised the weird shit we've found

> On Jul 21, 2015, at 23:17, David Lawrence <notifications@github.com> wrote:
> 
> In integration-cli/trust_server.go:
> 
> > +		return fmt.Errorf("notary ping replied with an unexpected status code %d", resp.StatusCode)
> > +	}
> > +	return nil
> > +}
> > +
> > +func (r *testNotary) Close() {
> > +	r.cmd.Process.Kill()
> > +	os.RemoveAll(r.dir)
> > +}
> > +
> > +func (s *DockerTrustSuite) trustedCmd(cmd *exec.Cmd) {
> > +	env := []string{
> > +		"DOCKER_TRUST=1",
> > +		fmt.Sprintf("DOCKER_TRUST_SERVER=%s", s.not.address()),
> > +		"DOCKER_TRUST_ROOT_PASSPHRASE=12345678",
> > +		"DOCKER_TRUST_TARGET_PASSPHRASE=12345678",
> Should have separate fuzzing test for that. Tests should fail for deterministic reasons. If a fuzzing test ever causes a failure, I'd expect the problematic value to be captured and a new test added specifically for that value to ensure no regressions in the future.
> 
> —
> Reply to this email directly or view it on GitHub.
>

endophage replied:
I don't disagree. Like I said though, tests should fail for deterministic reasons. So if fuzzing finds input that breaks a test, a new test should be written, after fixing the problem, to ensure no regressions occur down the road.

jfrazelle commented on registry/config.go:
```diff
@@ -32,6 +32,7 @@ const (
 	INDEXNAME      = "docker.io"
+	NOTARYSERVER   = "https://notary.docker.io"
 
 	// INDEXSERVER = "https://registry-stage.hub.docker.com/v1/"
```
ew can we remove this

endophage replied:
That's in master, you don't need to ask us for permissions to remove it :-)

jfrazelle replied:
Ya but since you are here...

> On Jul 21, 2015, at 23:20, David Lawrence <notifications@github.com> wrote:
> 
> In registry/config.go:
> 
> > @@ -32,6 +32,7 @@ const (
> >  	REGISTRYSERVER = DEFAULT_V2_REGISTRY
> >  	INDEXSERVER    = DEFAULT_V1_REGISTRY + "/v1/"
> >  	INDEXNAME      = "docker.io"
> > +	NOTARYSERVER   = "https://notary.docker.io"
> >  
> >  	// INDEXSERVER = "https://registry-stage.hub.docker.com/v1/"
> That's in master, you don't need to ask us for permissions to remove it :-)
> 
> —
> Reply to this email directly or view it on GitHub.
>

jfrazelle replied:
I shouldn't have to open a new PR to just remove this line..

endophage replied:
British humour doesn't seem to translate very well over github comments :-P

jfrazelle replied:
I'm really bad at British humor, and I consider myself to have dry humor haha my talks are always like crickets over there

> On Jul 22, 2015, at 11:33, David Lawrence <notifications@github.com> wrote:
> 
> In registry/config.go:
> 
> > @@ -32,6 +32,7 @@ const (
> >  	REGISTRYSERVER = DEFAULT_V2_REGISTRY
> >  	INDEXSERVER    = DEFAULT_V1_REGISTRY + "/v1/"
> >  	INDEXNAME      = "docker.io"
> > +	NOTARYSERVER   = "https://notary.docker.io"
> >  
> >  	// INDEXSERVER = "https://registry-stage.hub.docker.com/v1/"
> British humour doesn't seem to translate very well over github comments :-P
> 
> —
> Reply to this email directly or view it on GitHub.
>

jfrazelle commented on registry/registry.go:
```diff
@@ -61,6 +65,54 @@ func hasFile(files []os.FileInfo, name string) bool {
+		}
+		if strings.HasSuffix(f.Name(), ".key") {
+			keyName := f.Name()
+			certName := keyName[:len(keyName)-4] + ".cert"
```
couldnt this panic

jfrazelle replied:
hmmm i guess this was just moved from before...

endophage replied:
Same as above. Worst case, because we know ".key" will be present, we'll do keyName[:0]

calavera replied:
could this be `strings.Replace(f.Name(), ".key", ".cert", 1)` ? I'm not sure what's safer.

aaronlehmann replied:
That's not quite the same because it would replace the first instance of `.key`, but what about `strings.TrimSuffix(f.Name(), ".key") + ".cert"` ?

calavera replied:
agreed @aaronlehmann, thanks!

jfrazelle commented on registry/registry.go:
```diff
@@ -61,6 +65,54 @@ func hasFile(files []os.FileInfo, name string) bool {
+		}
+		if strings.HasSuffix(f.Name(), ".cert") {
+			certName := f.Name()
+			keyName := certName[:len(certName)-5] + ".key"
```
couldnt this panic as well

endophage replied:
No, because we jut checked that the suffix ".cert" is present which is 5 characters, so worst case this will be certName[:0] which will always return empty string.

NathanMcCauley committed:
Add trust tests for Docker create, run, push, and pull

Created date util function

Signed-off-by: Nathan McCauley <nathan.mccauley@docker.com>

calavera commented:
unfortunately, this is going to need a rebase :8ball:

jfrazelle commented:
ping @tianon for dockerfile / vendor changes and anything ;)

diogomonica committed:
Add test for incorrect nonroot passphrase

Fix failing tests for create, push, and pull

Signed-off-by: Diogo Monica <diogo@docker.com>

NathanMcCauley committed:
Add tests simulating an evil server

Signed-off-by: Nathan McCauley <nathan.mccauley@docker.com>

jfrazelle commented:
WOOHOOOOOOOOOO TESTSSSSSSSS <3

diogomonica committed:
Added tests for expired snapshots and timestamps

Signed-off-by: Diogo Monica <diogo@docker.com>

jfrazelle commented:
LGTM

jfrazelle commented:
(may need to squash some commits before merge tho)

jfrazelle commented:
LGTM 

(only if https://github.com/docker/docker/pull/14546/files#r35289269 is fixed and some commits are squashed before merge)

Thanks for the tests, seriously gives some peace of mind.

dmcgowan commented:
Added error when trying to create a notary client without an https endpoint. A user should only come across this if they explicitly set notary to use an http endpoint.

I also squashed down a few of my own commits which were functionally similar. I would prefer not to squash down farther since the commits are cleanly separated now by authorship or functionality.

dmcgowan committed:
Vendor notary

Signed-off-by: Derek McGowan <derek@mcgstyle.net> (github: dmcgowan)

dmcgowan committed:
Add build integration cli tests

Clean up tests to remove duplicate code

Add tests which run pull and create in an isolated configuration directory.
Add build test for untrusted tag

Signed-off-by: Derek McGowan <derek@mcgstyle.net> (github: dmcgowan)

calavera commented on api/client/build.go:
```diff
@@ -62,147 +64,109 @@ func (cli *DockerCli) CmdBuild(args ...string) error {
+		Compression:     archive.Uncompressed,
+		ExcludePatterns: excludes,
+		IncludeFiles:    includes,
+	}); err != nil {
```
can we split this big `if` into `asign+if`? it reads kinda weird.

calavera commented on api/client/build.go:
```diff
@@ -62,147 +64,109 @@ func (cli *DockerCli) CmdBuild(args ...string) error {
-		})
-	}
+	sf := streamformatter.NewStreamFormatter()
+	var body io.Reader = progressreader.New(progressreader.Config{
```
this can also be `body := progressreader...`

dmcgowan replied:
My understanding is that was done this way to explicitly cast to `io.Reader`

calavera replied:
oh, I see, cool :+1:

jlhawn replied:
> My understanding is that was done this way to explicitly cast to io.Reader

yes

calavera commented on api/client/build.go:
```diff
@@ -315,3 +277,294 @@ func (cli *DockerCli) CmdBuild(args ...string) error {
+func replaceDockerfileTarWrapper(inputTarStream io.ReadCloser, newDockerfile *trustedDockerfile, dockerfileName string) io.ReadCloser {
+	pipeReader, pipeWriter := io.Pipe()
+
+	go func() {
```
how do we control that this goroutine doesn't leak data? It looks like we could send the tar context to the server before this finishes.

jlhawn replied:
> It looks like we could send the tar context to the server before this finishes.

This goroutine is meant to run *while* it's being sent to the server. Reads from the pipe will block until data is written by the writer in this goroutine so, yes, the goroutine will not complete until the request body is completely sent to the daemon.

calavera replied:
From conversation with @jlhawn in IRC:

```
[11:00:30]  <+jlhawn>	that goroutine is meant to run while we send archive to the server
[11:00:49]  <+jlhawn>	we stream the tar through the pipe in this goroutine
[11:01:18]  <+jlhawn>	and the pipe is blocking, so it wont finish until the archive is completely uploaded to the server
```

calavera commented:
I have a small concern about a goroutine out of control, but I :heart: all the thorough testing in this PR. Thank you.

calavera commented:
I have two minor code style changes, but otherwise LGTM.

icecrime commented:
Give a minute so that I can give a final look, and we'll merge it if docs maintainers agree so that we can cut the 1.8.0-rc1.

icecrime commented on api/client/trust.go:
```diff
@@ -0,0 +1,445 @@
+	"github.com/endophage/gotuf/data"
+)
+
+var untrusted bool
```
:cry:

icecrime commented on api/client/trust.go:
```diff
@@ -0,0 +1,445 @@
+	// TODO(dmcgowan): false if name is hex image id
+	// TODO(dmcgowan): whitelist specific names to be untrusted
+	// TODO(dmcgowan): canonicalize name before checking whitelist
+	return !untrusted
```
As soon as 1.8.0 is out and we announce image signature, people will want to look at its implementation. This is not giving a good image: are those TODO features/optims that will make it in the future? Would they be better filed as issues?

I'm fine documenting potential future improvements as comments in the code, but in that let's explain clearly what they mean, or remove them altogether. Here it's vague enough to become suspicious.

icecrime commented on api/client/trust.go:
```diff
@@ -0,0 +1,445 @@
+		if alias == "snapshot" && snapshotEnv != "" {
+			return snapshotEnv, numAttempts > 1, nil
+		}
+		return baseRetriever(keyName, alias, createNew, numAttempts)
```
Cosmetic, but reduces duplication?

```
env := map[string]string {
    "root":    os.Getenv("DOCKER_TRUST_ROOT_PASSPHRASE"),
    "target":  os.Getenv("DOCKER_TRUST_TARGET_PASSPHRASE"),
    "snapshot: os.Getenv("DOCKER_TRUST_SNAPSHOT_PASSPHRASE"),
}
return func(keyName string, alias string, createNew bool, numAttempts int) (string, bool, error) {
    if v, := env[alias]; v != "" {
        return v, numAttempts > 1, nil
    }
    return baseRetriever(keyName, alias, createNew, numAttempts)
}
```

icecrime commented on api/client/trust.go:
```diff
@@ -0,0 +1,445 @@
+	snapshotEnv := os.Getenv("DOCKER_TRUST_SNAPSHOT_PASSPHRASE")
+
+	return func(keyName string, alias string, createNew bool, numAttempts int) (string, bool, error) {
+		if alias == "root" && rootEnv != "" {
```
Cosmetic, but reduces duplication?

```
env := map[string]string {
    "root":    os.Getenv("DOCKER_TRUST_ROOT_PASSPHRASE"),
    "targets": os.Getenv("DOCKER_TRUST_TARGET_PASSPHRASE"),
    "snapshot: os.Getenv("DOCKER_TRUST_SNAPSHOT_PASSPHRASE"),
}
return func(keyName string, alias string, createNew bool, numAttempts int) (string, bool, error) {
    if v, := env[alias]; v != "" {
        return v, numAttempts > 1, nil
    }
    return baseRetriever(keyName, alias, createNew, numAttempts)
}
```

dmcgowan replied:
agreed, this is cleaner I will update

icecrime commented on docs/reference/commandline/cli.md:
```diff
@@ -46,6 +46,8 @@ by the `docker` command line:
 * `DOCKER_RAMDISK` If set this will disable 'pivot_root'.
 * `DOCKER_TLS_VERIFY` When set Docker uses TLS and verifies the remote.
+* `DOCKER_TRUST` When set Docker uses notary to sign and verify images.
+  Equates to `--untrusted=false` for build, create, pull, push, run.
```
What about `DOCKER_TRUST_SERVER`?

dmcgowan replied:
Customizing the `DOCKER_TRUST_SERVER` is something we purposely left out of the documentation. This was only added for testing and is not currently a work flow we want to document and market. Do you think despite this it should be documented here? Same could be said of the `*_PASSPHRASE` environment variables.

tiborvass replied:
I don't think we should document these.

calavera replied:
I agree that we shouldn't publicize that variable for now. Let's wait until trusted is the default behavior.

icecrime commented on hack/vendor.sh:
```diff
@@ -38,6 +38,11 @@ clone git github.com/hashicorp/consul v0.5.2
 
+clone git github.com/docker/notary 9926ead788e6a5657de62560a4fa25c0a771a2d8
+clone git github.com/endophage/gotuf 374908abc8af7e953a2813c5c2b3944ab625ca68
+clone git github.com/tent/canonical-json-go 96e4ba3a7613a1216cbd1badca4efe382adea337
```
I said it already, but I really hope that in the future we could use our own implementation of this, with a properly documented set of difference with the standard golang package, and a way to track what version of the standard golang package it's based on (what's gonna happen when we move on to Go 1.5 here?).

icecrime replied:
Example of what I think would be a better quality dep: a repo holding a generator script which git clones the golang standard package with a fixed hash/tag and applies a diff. I believe it would make the dependency much more auditable: right now it's kind of a black box.

jfrazelle replied:
omg yes, we can do that! i mean i dont want to but if necessary i will

On Thu, Jul 23, 2015 at 2:24 PM, Arnaud Porterie <notifications@github.com>
wrote:

> In hack/vendor.sh
> <https://github.com/docker/docker/pull/14546#discussion_r35373168>:
>
> > @@ -38,6 +38,11 @@ clone git github.com/hashicorp/consul v0.5.2
> >  clone git github.com/docker/distribution 419bbc2da637d9b2a812be78ef8436df7caac70d
> >  clone git github.com/vbatts/tar-split v0.9.4
> >
> > +clone git github.com/docker/notary 9926ead788e6a5657de62560a4fa25c0a771a2d8
> > +clone git github.com/endophage/gotuf 374908abc8af7e953a2813c5c2b3944ab625ca68
> > +clone git github.com/tent/canonical-json-go 96e4ba3a7613a1216cbd1badca4efe382adea337
>
> Example of what I think would be a better quality dep: a repo holding a
> generator script which git clones the golang standard package with a fixed
> hash/tag and applies a diff. I believe it would make the dependency much
> more auditable: right now it's kind of a black box.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/docker/docker/pull/14546/files#r35373168>.
>

jfrazelle commented:
before merge I would LOVVVVVEEEEEE it if this https://github.com/tent/canonical-json-go was updated to at least be rebased on the current package upstream.... its so old and gross

icecrime commented on pkg/ansiescape/split.go:
```diff
@@ -0,0 +1,89 @@
+import "bytes"
+
+// dropCR drops a terminal \r from the data.
+func dropCR(data []byte) []byte {
```
Aren't you better off using [`bytes.Trim`](http://golang.org/pkg/bytes/#Trim) here?

dmcgowan replied:
The code I was referencing in the golang codebase did not use trim either, I was not sure if using trim would cause a copy of the slice. For reference http://golang.org/src/bufio/scan.go?s=9746:9824#L277

dmcgowan replied:
I could see how this could be wrong in this case if there is both a leading and trailing `\r`, the original implementation only allowed to drop a single trailing value.

dmcgowan replied:
Nevermind I misread my own code, it will drop both since it does not immediately return

icecrime commented on pkg/ansiescape/split.go:
```diff
@@ -0,0 +1,89 @@
+
+import "bytes"
+
+// dropCR drops a terminal \r from the data.
```
Lie! It's removing leading `\r` too :cop:

dmcgowan replied:
Good call, can you guess what the original implementation did :grinning:

icecrime commented on registry/registry.go:
```diff
@@ -61,6 +65,54 @@ func hasFile(files []os.FileInfo, name string) bool {
+	for _, f := range fs {
+		if strings.HasSuffix(f.Name(), ".crt") {
+			if tlsConfig.RootCAs == nil {
+				// TODO(dmcgowan): Copy system pool
```
Should we rather file an issue and remove the comment?

dmcgowan replied:
This was just moved over so don't want to change the comment. An issue already exists at https://github.com/docker/docker/issues/12756

icecrime commented:
Something's wrong: why is the current changeset dropping `vbatts/tar-split` from vendoring?

icecrime commented on api/client/trust.go:
```diff
@@ -0,0 +1,445 @@
+		if err != nil {
+			// treat any other value as true
+			trusted = true
+		}
```
```
if t, err := strconv.ParseBool(e); t || err != nil {
    trusted = true
}
```

dmcgowan commented:
Fixing vendor commit, not sure why Vincent's code was filtered out :smile:

icecrime commented on api/client/build.go:
```diff
@@ -315,3 +278,294 @@ func (cli *DockerCli) CmdBuild(args ...string) error {
+			}
+			ref := registry.ParseReference(tag)
+
+			if !ref.HasDigest() && isTrusted() {
```
Mmm.. but if the ref is already a digest, aren't we incorrectly passing through here?

icecrime replied:
Ok so I guess we will just avoid replacing here, but it will still be caught as a `notaryError` later at pull time.

dmcgowan commented:
Updated vendoring to not removed unrelated tar-split code, but the problem appears to be with master and not my code. The `hack/vendor.sh` script will remove those files.

dmcgowan commented:
Rebased (probably wasn't needed since tar-split issue is in master) and addressed comments

icecrime commented:
Small UX concern: because of the client-side Dockerfile rewrite, the pulls issued from a build don't get a tag. Example:

```
$ cat Dockerfile
FROM icecrimetest/build:latest
$ docker build -t icecrimetest/foo .
[...]
$ docker images
REPOSITORY           TAG                 IMAGE ID            CREATED              VIRTUAL SIZE
icecrimetest/build   latest              b7a6e397d84c        About a minute ago   5.25 MB
icecrimetest/bar     <none>              31f630c65071        5 weeks ago          5.25 MB

```

icecrime commented:
When pulling with trust, you get an extra entry with `docker images` that holds the repository with a `<none>` tag:

```
root@c155b3e71952:/tmp/foo# docker pull --untrusted=false icecrimetest/build:1.0
Pull (1 of 1): icecrimetest/build:1.0@sha256:73322764b50cb1b3a20f8247bfd0b4b79a196bd436601955658d5f56ce91f3d5
sha256:73322764b50cb1b3a20f8247bfd0b4b79a196bd436601955658d5f56ce91f3d5: Pulling from icecrimetest/build

b7a6e397d84c: Pull complete 
31f630c65071: Already exists 
Digest: sha256:73322764b50cb1b3a20f8247bfd0b4b79a196bd436601955658d5f56ce91f3d5
Status: Downloaded newer image for icecrimetest/build@sha256:73322764b50cb1b3a20f8247bfd0b4b79a196bd436601955658d5f56ce91f3d5
Tagging icecrimetest/build@sha256:73322764b50cb1b3a20f8247bfd0b4b79a196bd436601955658d5f56ce91f3d5 as icecrimetest/build:1.0
root@c155b3e71952:/tmp/foo# docker images
REPOSITORY           TAG                 IMAGE ID            CREATED             VIRTUAL SIZE
icecrimetest/build   1.0                 b7a6e397d84c        15 minutes ago      5.25 MB
icecrimetest/build   <none>              b7a6e397d84c        15 minutes ago      5.25 MB
```

tiborvass commented:
@icecrime that's already the case for pull by digest, which is what this PR basically does (translates tags to digests via notary).

icecrime commented:
@tiborvass No: pull by digest doesn't tag *at all* (which is expected: he doesn't even know about the tag).

icecrime commented:
This one fails but honestly that's just me being a b*, I don't think it's worth fixing:

```
$ cat Dockerfile
FROM ubu\
ntu
$ docker build -t boom .
unable to process Dockerfile: repository name component must match "[a-z0-9]+(?:[._-][a-z0-9]+)*"
```

dmcgowan commented:
@icecrime the lack of line continuation support has been noted by Josh. We both agreed it is enough of an edge case to ignore for 1.8. It is non-trivial to add but something we should support once the client side builder is implemented.

jfrazelle commented:
If it is non-trivial to add can we fix before release?

On Thu, Jul 23, 2015 at 5:01 PM, Derek McGowan <notifications@github.com>
wrote:

> @icecrime <https://github.com/icecrime> the lack of line continuation
> support has been noted by Josh. We both agreed it is enough of an edge case
> to ignore for 1.8. It is non-trivial to add but something we should support
> once the client side builder is implemented.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/docker/docker/pull/14546#issuecomment-124272578>.
>

icecrime commented:
A bit of a side effect when using `docker run` with trust enabled on a signed image:

```
$ docker run --untrusted=false -ti --rm icecrimetest/bar /bin/sh
$ docker ps -a
CONTAINER ID        IMAGE                                                                                      COMMAND             CREATED                  STATUS                  PORTS               NAMES
97ce21d9efb3        icecrimetest/bar@sha256:6a65b20394d35a782a560c56531bd74a05be016d9991cb1fcd9750524f31c40c   "/bin/sh"           Less than a second ago   Up Less than a second                       jovial_fermat
```

It would be better if the `IMAGE` here would be `icecrimetest/bar:latest` instead of the digest. I guess that's an easy change on the client side (send the container create with the tag rather than with the hash).

calavera commented:
@icecrime but the whole point of this PR is to create the container with the digest. We must send the digest to ensure the container is created with the trusted digest

icecrime commented:
@calavera That's not entirely true. The digest indirection is an implementation detail of the trust feature. I ask to run a container for a tagged image, I expect to tagged image to be used. Think of automated tools: I'm afraid it's going to be very disturbing to invoke run on `icecrimetest/bla` and see `icecrimetest@sha256:...` in the `ps`/`inspect` output.

calavera commented:
@icecrime the implementation detail is THE FEATURE. If you send the tag to the daemon nothing guarantees that you download the trusted digest.

icecrime commented:
@calavera But I don't want to know! I docker run `ubuntu:15.04`, I expect `ubuntu:15.04` to be the image name in my `ps` output and my container's `inspect`! What sense do I make of this hash? I don't even have an easy way to get a matching tag from it.

dmcgowan commented:
Added another integration test for building with an untrusted tag

endophage commented:
LGTM if the tests pass

diogomonica commented:
I understand @icecrime concerns. I think we should absolutely take a look at how people use this during 1.8. I don't think any of these are a blocker though.

LGTM

icecrime commented:
Of all things I noted, the extra tag is for me the most problematic one.

From a clean install, doing `DOCKER_TRUSTED=1 docker pull foo && docker rmi foo` and having a lingering `<none>` tag is unexpected, and could lead to some "image leak" on automated systems. But I'll go with the majority.

jfrazelle commented:
I do not like that either...

dmcgowan committed:
Rename to flags and environment variables to content trust

Update help line to allow 90 characters instead of 80

The trust flag pushes out the help description column wider, requiring more room to display help messages.

Signed-off-by: Derek McGowan <derek@mcgstyle.net> (github: dmcgowan)

icecrime commented:
Waiting on CI. Docs will come on followup PR (we want to be able to cut 1.8.0-rc1 branch).

icecrime commented:
Aaaaand it happened

```
----------------------------------------------------------------------
FAIL: docker_cli_help_test.go:14: DockerSuite.TestHelpTextVerify

docker_cli_help_test.go:137:
    c.Fatalf("Help for %q is too long(%d chars):\n%s", cmd,
        len(line), line)
... Error: Help for "build" is too long(86 chars):
  --cpu-period=0                  Limit the CPU CFS (Completely Fair Scheduler) period
```

dmcgowan commented:
Interesting it failed on "build" since it looks like we forgot to update that one

jfrazelle commented:
are u going to kill me if i ask if we can squash the last 4 commits into one

dmcgowan commented:
I would ask what the value is, I could easily squash the 3rd to last, the last one was left unsquash to easily backout (no emotional attachment), and the second to last I left untouched to make updates easier and the squash should be ugly to put each change in their proper place :smile:.

dmcgowan commented:
Bah ci failed anyway, squashed away 2 more and pushed

jfrazelle commented:
<3

jfrazelle commented:
started lxc build, bc why nottttttt :innocent:

icecrime commented:
Please ignore ping failures against `www.docker.io` in CI.

icecrime commented:
Ignoring failures on www.docker.io pings: we're good to go.

jlhawn commented:
:tada:

jlhawn commented:
so, should I submit https://github.com/dmcgowan/docker/pull/25/files against upstream then?

phemmer commented:
Just fyi, this PR is the cause of #15785
